Initial MongoDB setup for MSUnmerged.

todor-ivanov opened the pull request:
Fixes #10798 

#### Status
Ready

#### Description
This PR is providing the basic functionalities we need to be able to write the proper structure to MongoDB for the MSUnmerged micro service.

#### Is it backward compatible (if not, which system it affects?)
NO, once we have that merged we  need to have a MongoDB instance configured and running. It will also break several stuff related to the microservice work itself:
 * We may no longer rely on refreshing the timers from memory, since we are about to be preserving those in the database
 * We will keep accumulating garbage records in the fields  of the RSE object if we just update and never clean the RSE record at the database upon refresh of the RSE read on the RucioConMon side  

#### Related PRs
Here are the configuration commits related to the initial MongoDb setup [1]. We are using the MongoDB as a service instance so far and distinguishing between service instances by database names (fetched from service_config).

[1]
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/121/diffs
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/122/diffs

#### External dependencies / deployment changes
We need to propagate the correct passwords in the relevant password file (this is happening in a private mode with the cmsweb Team)

todor-ivanov committed:
Initial MongoDB setup.

Add database read and write logic to MSUnmergedRSE objects.

Convert between datatypes for pythonlike and Jsonlike objects.

Fix some misaligned variable names

Add full object write to database functionality in the MSUnmergedRSE object.

Review follow up.

Rename flag dumpRSE && Read database name from msConfig.

todor-ivanov commented:
Hi @amaltaro, since we just introduced MongoDB into the picture, those unit tests for MSUnmerged started to fail, because the machine to run the tests at Jenkins has no access to any MongoDB instance, neither as a service nor local.  As of code-wise the changes are now completed and well tested manually with the already existing MongoDB as a Service instance from my virtual environment and the results were all fine. 

So I'd suggest you may start you review of the code until I figure out what to do with the broken unit tests.

BTW any ideas on how to simulate the database or to connect to the already existing one without exposing any usernames or passwords here are very welcome.

todor-ivanov commented:
And just for refreshing, we had the very same problem with unit tests for MSOutput and it is still not resolved. The GH issue related to the problem is this one: https://github.com/dmwm/WMCore/issues/9789

There are few options here.:
 * The simplest one - we can install MongoDB at the machine running the tests - it will always work since all services by default point to local host.
 * Use libraries as [mongomock]( https://github.com/mongomock/mongomock) for in-memory emulation of MongoDB. Or a test library relying on the above one and available through `pipy` [pytest-mongodb](https://pypi.org/project/pytest-mongodb/). Still need to figure out if this would be just a database emulation or a full server emulation. Because we will actually need the full server emulation to simulate  the communication protocol itself. Reason is, the MongoDB connection for our objects happen at init time [1] so the server client communication is unavoidable.
 * The third option would be working just for database emulation with some simple code like: 
 
```
@mock.patch("pymongo.collection.Collection.find")
def test_name(self, mock_find):
    mock_find.return_value = MongoDummybject(return_data)
```    

**NOTE:** example taken from [stackoverflow](https://stackoverflow.com/a/68124776)

[1]

https://github.com/dmwm/WMCore/blob/dd3d3252b2f545ef84d65573d3e69216da06ad4a/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L121

todor-ivanov commented:
Hi @amaltaro as we spoke earlier today I did add the functionality to write the whole RSE object to the database. And I did keep the PR split in two separate commits, but the first one has some broken database write logic, which I've  fixed in the second one. So I'd suggest we go with them both and not looking back. 

The databse `read && write` operations are now well tested for every possible case both manually and combined with the calls from MSUnmerged code itself. What I didn't do so far, is changing the flag name for triggering the behavior (`dumpRSE`). This will require some changes in the config repositories as well. I'd better proceed with the whole change in a separate commit, which may also include the needed configuration changes to add a simple connection to MongoDB as a service.

amaltaro reviewed (changes requested):
Todor, please find some questions and requests along the code.

I also have a general comment/question regarding these proposed changes. I think we discussed in the past week that a failure in the pipeline should not trigger a RSE document update operation in MongoDB. Giving it a second thought, maybe it would actually be better to have this information getting persisted in MongoDB and retrievable via REST API in the near future. Of course, only when cleanup is actually required against a given RSE. What do you think?

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -101,6 +105,22 @@ def __init__(self, msConfig, logger=None):
+        msUnmergedIndex = IndexModel('name', unique=True)
+        msUnmergedDBConfig = {
+            'database': 'msUnmergedDB',
+            'server': self.msConfig['mongoDBUrl'],
```
not yet. They are to come with the PR in the service configuration repository together with the variable renaming as I mentioned in one of the comments.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -101,6 +105,22 @@ def __init__(self, msConfig, logger=None):
+
+        msUnmergedIndex = IndexModel('name', unique=True)
+        msUnmergedDBConfig = {
+            'database': 'msUnmergedDB',
```
I am not sure how these MongoDB databases would be managed in the future. So far we  are in charge of creating them and it happens automatically at a connection session creation (this is how we have our MongoDB wrapper build) . In order to change the database name we'll need to change the name here and also a variable name in the code, which may happen at any time. Lets first see how and who, besides us, would use the service and what, if any, naming schema would be applied in general.

amaltaro replied:
Ahh, I forgot this detail about the attribute name and the database name. Thanks

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+                        included in the projection ergo saved into the database.
+        :return:        A dictionary type MongoDB Projection
+        """
+        if dumpRSE:
```
Nice touch :). I'll make the change.

todor-ivanov replied:
done

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+    def writeRSEToMongoDB(self, collection, dumpRSE=False, overwrite=False):
+        """
+        A method to write/update the RSE at the Database from the current object.
+        :param collection: The MonogoDB collection to write on
```
BTW  `self.purgeRSEAtMongoDB` needs to be created, otherwise we will keep accumulating old records upon refresh of the RSE record from RucioConMon. I just forgot to put it in this commit. I was about to create and call the method in the PR related to `stateful` vs `stateless` service migration.

amaltaro replied:
Okay, we better leave this one (and a couple others) unresolved such that we can reference to this in the upcoming developments.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+        :param collection: The MonogoDB collection to read from
+        :return:           True if read and update were both successful, False otherwise.
+        """
+        # NOTE: The fields to be manipulated are only those which are compatible
```
Kind of. I'll remove it.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+    def readRSEFromMongoDB(self, collection):
+        """
+        A method to read the RSE object from Database and update it's fields.
+        :param collection: The MonogoDB collection to read from
```
done

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -690,6 +717,26 @@ def resetCounters(self, rseName=None, plineName=None):
+    def uploadRSEToMongoDB(self, rse, dumpRSE=False, overwrite=True):
+        """
+        Updates the record for an RSE at MongoDB
+        :param rse: The RSE object to work on
```
done

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -690,6 +717,26 @@ def resetCounters(self, rseName=None, plineName=None):
+        :return:    rse
+        """
+        if not dumpRSE:
+            dumpRSE = self.msConfig['dumpRSE']
```
Yes, I am fine with the function not having an additional step for setting a default behavior and relying only on its default parameters. I'll remove it.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+        for field in mongoProjection:
+            if mongoProjection[field]:
+                # convert back the strictly pythonic `set`s && `generators` to Json compatible lists
+                if field == 'dirs':
```
done

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:
```diff
@@ -118,14 +138,16 @@ def __init__(self, msConfig, logger=None):
         self.plineUnmerged = Pipeline(name=pName,
-                                      funcLine=[Functor(self.updateRSETimestamps, start=True, end=False),
+                                      funcLine=[Functor(self.getRSEFromMongoDB),
+                                                Functor(self.updateRSETimestamps, start=True, end=False),
```
Yes this rework needs to be done. And I was trying to mention this during the meeting earlier today. And not only that we are also risking to start accumulating old records in the fields, if we do not purge the record in the database once in a while (i.e. when RucioConMon refreshes the reading). I was planing all those to go in the next issue.

todor-ivanov commented on src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py:
```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+                    updateFields[field] = self[field]
+
+        updateOps = {'$set': updateFields}
+        if overwrite:
```
Almost...

This:
> However, when overwrite=False, a new document does not get created in case it does not exist (it does here because upset=True...).

Is not what is meant in the logic here. `upsert` will always take precedence no mater what the flag of `overwrite` is. `upsert` tells if a missing document is about to be created. `overwrite` relates only to an already existing document. So if a document is missing and the call is telling: "please create the document" it will be created no matter what the other flags stands.
 
And also the update operation is supposed to be working on fields, while the replace operation is supposed  to swap the whole document.

todor-ivanov commented:
@amaltaro Alan please take another look. I think I have addressed most of your comments. Some had to be left for the other issue. And I am currently working  on the configuration PR separately. I'll push that  one soon too. Thanks!

amaltaro reviewed (commented):
Todor, can you please:
* squash these commits
* make the required configuration changes
* and update the initial description with the required dependencies
?

Once that's done, it's good to go. Thanks!

todor-ivanov commented:
Hi @amaltaro I just did all requested :
 * Squashed all the commits
 * Made the configuration changes
 * Properly linked in the comments 
 * Variable renamed

Could you please take another look 

Thanks in advance!

amaltaro reviewed (approved):
Thanks Todor!

amaltaro commented:
Even though we want to stop using our VMs for central services tests, we still do not have a full k8s dev/test setup. So here is a fix that should have gone with this PR: https://github.com/dmwm/deployment/pull/1115
